File: problems/009/009.rb
def f
  perimeter = 1000

  ((perimeter / 3 + 1)..(perimeter / 2)).each do |c|
    (((perimeter - c) / 2 + 1)..c).each do |b|
      a = perimeter - b - c

      return a * b * c if a * a + b * b == c * c
    end
  end
end

case ARGV.length
when 0
  puts f
when 1
  iters = ARGV.first.to_i

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  (1..iters).each do ||
    f
  end
  end_ = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  elapsed = ((end_ - start) * 1e9).to_i

  puts elapsed
end

File: problems/009/009.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define PERIMETER 1000

__attribute__((always_inline))
inline uint64_t f() {
  uint64_t a, b, c;

  for (c = PERIMETER / 3 + 1; c < PERIMETER / 2; c++)
    for (b = (PERIMETER - c) / 2 + 1; b < c; b++) {
      a = PERIMETER - b - c;

      if (a * a + b * b == c * c)
        return a * b * c;
    }
}

int64_t to_ns(struct timespec ts) {
  return (uint64_t)ts.tv_sec * 1000000000lu + (uint64_t)ts.tv_nsec;
}

int main(int argc, char *argv[]) {
  if (argc == 1) {
    printf("%lu\n", f());
  } else if (argc == 2) {
    struct timespec start = {0, 0}, end = {0, 0};
    uint64_t i = 0;
    uint64_t iters = atol(argv[1]);

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (; i < iters; i++) {
      f();
      asm("");
    }
    clock_gettime(CLOCK_MONOTONIC, &end);

    printf("%lu", to_ns(end) - to_ns(start));
  } else {
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}
